File: src/main/java/me/shedaniel/linkie/commands/YarnFieldCommand.java
package me.shedaniel.linkie.commands;

import me.shedaniel.linkie.CommandBase;
import me.shedaniel.linkie.InvalidUsageException;
import me.shedaniel.linkie.LinkieBot;
import me.shedaniel.linkie.yarn.MappingsData;
import me.shedaniel.linkie.yarn.YarnField;
import me.shedaniel.linkie.yarn.YarnManager;
import net.fabricmc.mappings.EntryTriple;
import org.javacord.api.entity.message.MessageAuthor;
import org.javacord.api.entity.message.embed.EmbedBuilder;
import org.javacord.api.event.message.MessageCreateEvent;

import java.awt.*;
import java.util.List;
import java.util.*;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import static me.shedaniel.linkie.LinkieBot.contains;

public class YarnFieldCommand implements CommandBase {
    @Override
    public void execute(ScheduledExecutorService service, MessageCreateEvent event, MessageAuthor author, String cmd, String[] args) {
        if (YarnManager.updating)
            throw new RuntimeException("Yarn is being downloaded, please wait for around 10 seconds");
        if (args.length != 1 && args.length != 2)
            throw new InvalidUsageException("+" + cmd + " <search> [page]");
        event.getChannel().sendMessage(new EmbedBuilder().setTitle("Loading...").setFooter("Requested by " + author.getDiscriminatedName(), author.getAvatar()).setTimestampToNow()).whenComplete((message1, throwable) -> {
            try {
                int page = args.length > 1 ? Integer.parseInt(args[1]) - 1 : 0;
                String name = args[0].indexOf('.') > -1 ? args[0].split("\\.")[1] : args[0];
                String clazz = args[0].indexOf('.') > -1 ? args[0].split("\\.")[0] : null;
                String low = name.toLowerCase(Locale.ROOT);
                String lowClazz = clazz != null ? clazz.toLowerCase(Locale.ROOT) : null;
                List<YarnField> files = new ArrayList<>();
                YarnManager.r1_2_5.tiny.getFieldEntries().forEach(fieldEntry -> {
                    EntryTriple intermediary = fieldEntry.get("intermediary");
                    EntryTriple server = fieldEntry.get("server");
                    EntryTriple client = fieldEntry.get("client");
                    if (clazz == null || contains(getLast(intermediary.getOwner().toLowerCase(Locale.ROOT)), lowClazz) || (server != null && contains(getLast(server.getOwner().toLowerCase(Locale.ROOT)), lowClazz)) || (client != null && contains(getLast(client.getOwner().toLowerCase(Locale.ROOT)), lowClazz))) {
                        if (contains(getLast(intermediary.getName()).toLowerCase(Locale.ROOT), low) || (server != null && contains(server.getName(), low)) || (client != null && contains(client.getName(), low)))
                            files.add(new YarnField(intermediary, server, client));
                    }
                });
                YarnManager.r1_2_5.mappings.forEach(mappingsFile -> {
                    boolean matchClass = clazz == null || contains(getLast(mappingsFile.getObfClass().toLowerCase(Locale.ROOT)), lowClazz) || contains(getLast(mappingsFile.getDeobfClass().toLowerCase(Locale.ROOT)), lowClazz);
                    if (matchClass)
                        for(MappingsData.FieldMappings fieldMapping : mappingsFile.getFieldMappings())
                            if (contains(fieldMapping.getObfName().toLowerCase(Locale.ROOT), low) || contains(fieldMapping.getDeobfName().toLowerCase(Locale.ROOT), low))
                                files.add(new YarnField(new EntryTriple(mappingsFile.getObfClass(), fieldMapping.getObfName(), fieldMapping.getDesc()), new EntryTriple(mappingsFile.getDeobfClass(), fieldMapping.getDeobfName(), fieldMapping.getDesc())));
                });
                files.forEach(yarnField -> {
                    if (yarnField.incomplete())
                        if (yarnField.needObf()) {
                            YarnManager.r1_2_5.tiny.getFieldEntries().stream().filter(fieldEntry -> {
                                return getLast(fieldEntry.get("intermediary").getName()).equalsIgnoreCase(getLast(yarnField.getIntermediary().getName()));
                            }).findAny().ifPresent(classEntry -> {
                                yarnField.setClient(classEntry.get("client"));
                                yarnField.setServer(classEntry.get("server"));
                            });
                        } else if (yarnField.needMapped()) {
                            for(MappingsData.MappingsFile file : YarnManager.r1_2_5.mappings.stream().filter(mappingsFile -> {
                                return getLast(mappingsFile.getObfClass()).equalsIgnoreCase(getLast(yarnField.getIntermediary().getOwner()));
                            }).collect(Collectors.toList())) {
                                for(MappingsData.FieldMappings fieldMapping : file.getFieldMappings()) {
                                    if (fieldMapping.getObfName().equalsIgnoreCase(yarnField.getIntermediary().getName())) {
                                        yarnField.setMapped(new EntryTriple(file.getDeobfClass(), fieldMapping.getDeobfName(), fieldMapping.getDesc()));
                                        break;
                                    }
                                }
                            }
                        }
                });
                if (files.isEmpty())
                    throw new NullPointerException("Match fields not found!");
                files.sort(Comparator.comparingDouble(value -> similarity(get(value, getLast(low)), getLast(low))));
                Collections.reverse(files);
                EmbedBuilder builder = new EmbedBuilder().setTitle("List of Yarn Mappings (Page " + (page + 1) + "/" + (int) Math.ceil(files.size() / 5d) + ")").setFooter("Requested by " + author.getDiscriminatedName(), author.getAvatar()).setTimestampToNow();
                final String[] desc = {""};
                files.stream().skip(5 * page).limit(5).map(yarnField -> {
                    if (yarnField.needObf())
                        return null;
                    String obf = "client=" + (yarnField.getClient() == null ? "null" : yarnField.getClient().getName()) + ",server=" + (yarnField.getServer() == null ? "null" : yarnField.getServer().getName());
                    String main = yarnField.getMapped() != null ? yarnField.getMapped().getOwner() + "." + yarnField.getMapped().getName() : yarnField.getIntermediary().getOwner() + "." + yarnField.getIntermediary().getName();
                    return "**MC 1.2.5: " + main + "**\n__Name__: " + obf + " => `" + yarnField.getIntermediary().getName() + "`" + (yarnField.getMapped() != null ? " => `" + yarnField.getMapped().getName() + "`" : "") + "\n__Type__: `" + nameType(mapDesc(yarnField.getIntermediary().getDesc())).replace('/', '.') + "`\n__Mixin Target__: `" + turnDesc(yarnField.getMapped() != null ? yarnField.getMapped().getOwner() : yarnField.getIntermediary().getOwner()) + (yarnField.getMapped() != null ? yarnField.getMapped().getName() : yarnField.getIntermediary().getName()) + ":" + mapDesc(yarnField.getIntermediary().getDesc()) + "`";
                }).filter(Objects::nonNull).forEach(s -> {
                    if (desc[0].length() + s.length() > 1990)
                        return;
                    if (!desc[0].isEmpty())
                        desc[0] += "\n\n";
                    desc[0] += s;
                });
                builder.setDescription(desc[0].substring(0, Math.min(desc[0].length(), 2000)));
                message1.edit(builder);
                int finalPage[] = {page};
                if (message1.isServerMessage())
                    message1.removeAllReactions().get();
                message1.addReactions("⬅", "❌", "➡").thenRun(() -> {
                    message1.addReactionAddListener(reactionAddEvent -> {
                        try {
                            if (!reactionAddEvent.getUser().getDiscriminatedName().equals(LinkieBot.getApi().getYourself().getDiscriminatedName()) && !reactionAddEvent.getUser().getDiscriminatedName().equals(author.getDiscriminatedName())) {
                                reactionAddEvent.removeReaction();
                                return;
                            }
                            if (reactionAddEvent.getUser().getDiscriminatedName().equals(author.getDiscriminatedName()))
                                if (reactionAddEvent.getEmoji().equalsEmoji("❌")) {
                                    reactionAddEvent.deleteMessage();
                                } else if (reactionAddEvent.getEmoji().equalsEmoji("⬅")) {
                                    reactionAddEvent.removeReaction();
                                    if (finalPage[0] > 0) {
                                        finalPage[0]--;
                                        EmbedBuilder builder1 = new EmbedBuilder().setTitle("List of Yarn Mappings (Page " + (finalPage[0] + 1) + "/" + (int) Math.ceil(files.size() / 5d) + ")").setFooter("Requested by " + author.getDiscriminatedName(), author.getAvatar()).setTimestampToNow();
                                        final String[] desc1 = {""};
                                        files.stream().skip(5 * finalPage[0]).limit(5).map(yarnField -> {
                                            if (yarnField.needObf())
                                                return null;
                                            String obf = "client=" + (yarnField.getClient() == null ? "null" : yarnField.getClient().getName()) + ",server=" + (yarnField.getServer() == null ? "null" : yarnField.getServer().getName());
                                            String main = yarnField.getMapped() != null ? yarnField.getMapped().getOwner() + "." + yarnField.getMapped().getName() : yarnField.getIntermediary().getOwner() + "." + yarnField.getIntermediary().getName();
                                            return "**MC 1.2.5: " + main + "**\n__Name__: " + obf + " => `" + yarnField.getIntermediary().getName() + "`" + (yarnField.getMapped() != null ? " => `" + yarnField.getMapped().getName() + "`" : "") + "\n__Type__: `" + nameType(mapDesc(yarnField.getIntermediary().getDesc())).replace('/', '.') + "`\n__Mixin Target__: `" + turnDesc(yarnField.getMapped() != null ? yarnField.getMapped().getOwner() : yarnField.getIntermediary().getOwner()) + (yarnField.getMapped() != null ? yarnField.getMapped().getName() : yarnField.getIntermediary().getName()) + ":" + mapDesc(yarnField.getIntermediary().getDesc()) + "`";
                                        }).filter(Objects::nonNull).forEach(s -> {
                                            if (desc1[0].length() + s.length() > 1990)
                                                return;
                                            if (!desc1[0].isEmpty())
                                                desc1[0] += "\n\n";
                                            desc1[0] += s;
                                        });
                                        builder1.setDescription(desc1[0].substring(0, Math.min(desc1[0].length(), 2000)));
                                        message1.edit(builder1);
                                    }
                                } else if (reactionAddEvent.getEmoji().equalsEmoji("➡")) {
                                    reactionAddEvent.removeReaction();
                                    if (finalPage[0] < (int) Math.ceil(files.size() / 5d) - 1) {
                                        finalPage[0]++;
                                        EmbedBuilder builder1 = new EmbedBuilder().setTitle("List of Yarn Mappings (Page " + (finalPage[0] + 1) + "/" + (int) Math.ceil(files.size() / 5d) + ")").setFooter("Requested by " + author.getDiscriminatedName(), author.getAvatar()).setTimestampToNow();
                                        final String[] desc1 = {""};
                                        files.stream().skip(5 * finalPage[0]).limit(5).map(yarnField -> {
                                            if (yarnField.needObf())
                                                return null;
                                            String obf = "client=" + (yarnField.getClient() == null ? "null" : yarnField.getClient().getName()) + ",server=" + (yarnField.getServer() == null ? "null" : yarnField.getServer().getName());
                                            String main = yarnField.getMapped() != null ? yarnField.getMapped().getOwner() + "." + yarnField.getMapped().getName() : yarnField.getIntermediary().getOwner() + "." + yarnField.getIntermediary().getName();
                                            return "**MC 1.2.5: " + main + "**\n__Name__: " + obf + " => `" + yarnField.getIntermediary().getName() + "`" + (yarnField.getMapped() != null ? " => `" + yarnField.getMapped().getName() + "`" : "") + "\n__Type__: `" + nameType(mapDesc(yarnField.getIntermediary().getDesc())).replace('/', '.') + "`\n__Mixin Target__: `" + turnDesc(yarnField.getMapped() != null ? yarnField.getMapped().getOwner() : yarnField.getIntermediary().getOwner()) + (yarnField.getMapped() != null ? yarnField.getMapped().getName() : yarnField.getIntermediary().getName()) + ":" + mapDesc(yarnField.getIntermediary().getDesc()) + "`";
                                        }).filter(Objects::nonNull).forEach(s -> {
                                            if (desc1[0].length() + s.length() > 1990)
                                                return;
                                            if (!desc1[0].isEmpty())
                                                desc1[0] += "\n\n";
                                            desc1[0] += s;
                                        });
                                        builder1.setDescription(desc1[0].substring(0, Math.min(desc1[0].length(), 2000)));
                                        message1.edit(builder1);
                                    }
                                }
                        } catch (Throwable throwable1) {
                            throwable1.printStackTrace();
                        }
                    }).removeAfter(30, TimeUnit.MINUTES);
                });
            } catch (Throwable throwable1) {
                throwable1.printStackTrace();
                message1.edit(new EmbedBuilder().setTitle("Linkie Error").setColor(Color.red).setFooter("Requested by " + event.getMessageAuthor().getDiscriminatedName(), event.getMessageAuthor().getAvatar()).addField("Error occurred while processing the command:", throwable1.getClass().getSimpleName() + ": " + throwable1.getLocalizedMessage()).setTimestampToNow());
            }
        });
    }
    
    private String turnDesc(String s) {
        return "L" + s + ";";
    }
    
    private String nameType(String desc) {
        if (desc.length() == 1 || (desc.length() == 2 && desc.charAt(0) == '[')) {
            char c = desc.length() == 1 ? desc.charAt(0) : desc.charAt(1);
            switch (c) {
                case 'Z':
                    return "boolean";
                case 'C':
                    return "char";
                case 'B':
                    return "byte";
                case 'S':
                    return "short";
                case 'I':
                    return "int";
                case 'F':
                    return "float";
                case 'J':
                    return "long";
                case 'D':
                    return "double";
            }
        }
        if (desc.startsWith("L"))
            return desc.substring(1, desc.length() - 1);
        if (desc.startsWith("[[L"))
            return desc.substring(3, desc.length() - 1);
        return desc;
    }
    
    private String mapDesc(String desc) {
        if (desc.startsWith("L")) {
            String substring = desc.substring(1, desc.length() - 1);
            Optional<MappingsData.MappingsFile> any = YarnManager.r1_2_5.mappings.stream().filter(mappingsFile -> mappingsFile.getObfClass().equals(substring)).findAny();
            if (any.isPresent())
                return "L" + any.get().getDeobfClass() + ";";
        }
        if (desc.startsWith("[[L")) {
            String substring = desc.substring(3, desc.length() - 1);
            Optional<MappingsData.MappingsFile> any = YarnManager.r1_2_5.mappings.stream().filter(mappingsFile -> mappingsFile.getObfClass().equals(substring)).findAny();
            if (any.isPresent())
                return "[[L" + any.get().getDeobfClass() + ";";
        }
        return desc;
    }
    
    public String get(YarnField yarnClass, String search) {
        String intermediary = getLast(yarnClass.getIntermediary().getName());
        if (contains(intermediary, search))
            return intermediary;
        if (yarnClass.getMapped() != null) {
            String mapped = getLast(yarnClass.getMapped().getName()).toLowerCase(Locale.ROOT);
            if (contains(mapped, search))
                return mapped;
        }
        if (yarnClass.getServer() != null) {
            String mapped = getLast(yarnClass.getServer().getName()).toLowerCase(Locale.ROOT);
            if (contains(mapped, search))
                return mapped;
        }
        if (yarnClass.getClient() != null) {
            String mapped = getLast(yarnClass.getClient().getName()).toLowerCase(Locale.ROOT);
            if (contains(mapped, search))
                return mapped;
        }
        return "jdkwjidhwudhuwihduhudwuhiwuhui";
    }
    
    public int editDistance(String s1, String s2) {
        s1 = s1.toLowerCase();
        s2 = s2.toLowerCase();
        
        int[] costs = new int[s2.length() + 1];
        for(int i = 0; i <= s1.length(); i++) {
            int lastValue = i;
            for(int j = 0; j <= s2.length(); j++) {
                if (i == 0)
                    costs[j] = j;
                else {
                    if (j > 0) {
                        int newValue = costs[j - 1];
                        if (s1.charAt(i - 1) != s2.charAt(j - 1))
                            newValue = Math.min(Math.min(newValue, lastValue), costs[j]) + 1;
                        costs[j - 1] = lastValue;
                        lastValue = newValue;
                    }
                }
            }
            if (i > 0)
                costs[s2.length()] = lastValue;
        }
        return costs[s2.length()];
    }
    
    public String getLast(String s) {
        if (s.indexOf('/') <= -1)
            return s;
        String[] s1Split = s.split("/");
        return s1Split[s1Split.length - 1];
    }
    
    public double similarity(String s1, String s2) {
        s1 = getLast(s1);
        s2 = getLast(s2);
        String longer = s1, shorter = s2;
        if (s1.length() < s2.length()) { // longer should always have greater length
            longer = s2;
            shorter = s1;
        }
        int longerLength = longer.length();
        if (longerLength == 0) {
            return 1.0; /* both strings are zero length */
        }
        return (longerLength - editDistance(longer, shorter)) / (double) longerLength;
    }
}

File: src/main/kotlin/me/shedaniel/linkie/commands/YarnFieldCommand.kt
package me.shedaniel.linkie.commands

import discord4j.core.`object`.entity.MessageChannel
import discord4j.core.`object`.entity.User
import discord4j.core.event.domain.message.MessageCreateEvent
import discord4j.core.event.domain.message.ReactionAddEvent
import discord4j.core.spec.EmbedCreateSpec
import me.shedaniel.linkie.*
import me.shedaniel.linkie.utils.dropAndTake
import me.shedaniel.linkie.utils.onlyClass
import me.shedaniel.linkie.utils.similarity
import me.shedaniel.linkie.utils.similarityOnNull
import java.time.Duration
import kotlin.math.ceil
import kotlin.math.min

object YarnFieldCommand : AYarnFieldCommand({ if (it.id.asLong() == 602959845842485258) "1.2.5" else latestYarn }) {
    override fun getName(): String? = "Yarn Field Command"
    override fun getDescription(): String? = "Query yarn fields."
}

object POMFFieldCommand : AYarnFieldCommand({ "b1.7.3" }) {
    override fun getName(): String? = "POMF Field Command"
    override fun getDescription(): String? = "Query pomf fields."
}

open class AYarnFieldCommand(private val defaultVersion: (MessageChannel) -> String) : CommandBase {
    override fun execute(event: MessageCreateEvent, user: User, cmd: String, args: Array<String>, channel: MessageChannel) {
        if (args.isEmpty())
            throw InvalidUsageException("!$cmd <search> [version]")

        val mappingsContainerGetter = tryLoadMappingContainer(args.last(), getMappingsContainer(defaultVersion.invoke(channel)))

        var searchTerm = args.joinToString(" ").replace('.', '/')
        if (mappingsContainerGetter.first == args.last()) {
            searchTerm = searchTerm.substring(0, searchTerm.lastIndexOf(' '))
        }
        if (searchTerm.contains(' '))
            throw InvalidUsageException("!$cmd <search> [version]")
        val hasClass = searchTerm.contains('/')
        val hasWildcard = (hasClass && searchTerm.substring(0, searchTerm.lastIndexOf('/')).onlyClass() == "*") || searchTerm.onlyClass('/') == "*"

        val message = channel.createEmbed {
            it.apply {
                setFooter("Requested by " + user.discriminatedName, user.avatarUrl)
                setTimestampToNow()
                var desc = "Searching up fields for **${mappingsContainerGetter.first}**."
                if (hasWildcard) desc += "\nCurrently using wildcards, might take a while."
                if (!mappingsContainerGetter.second) desc += "\nThis mappings version is not yet cached, might take some time to download."
                setDescription(desc)
            }
        }.block() ?: throw NullPointerException("Unknown Message!")

        try {
            val mappingsContainer = mappingsContainerGetter.third.invoke()

            val classes = mutableMapOf<Class, FindFieldMethod>()

            if (hasClass) {
                val clazzKey = searchTerm.substring(0, searchTerm.lastIndexOf('/')).onlyClass()
                if (clazzKey == "*") {
                    mappingsContainer.classes.forEach { classes[it] = FindFieldMethod.WILDCARD }
                } else {
                    mappingsContainer.classes.forEach { clazz ->
                        when {
                            clazz.intermediaryName.onlyClass().contains(clazzKey, true) -> classes[clazz] = FindFieldMethod.INTERMEDIARY
                            clazz.mappedName?.onlyClass()?.contains(clazzKey, true) == true -> classes[clazz] = FindFieldMethod.MAPPED
                            clazz.obfName.client?.onlyClass()?.contains(clazzKey, true) == true -> classes[clazz] = FindFieldMethod.OBF_CLIENT
                            clazz.obfName.server?.onlyClass()?.contains(clazzKey, true) == true -> classes[clazz] = FindFieldMethod.OBF_SERVER
                            clazz.obfName.merged?.onlyClass()?.contains(clazzKey, true) == true -> classes[clazz] = FindFieldMethod.OBF_MERGED
                        }
                    }
                }
            } else mappingsContainer.classes.forEach { classes[it] = FindFieldMethod.WILDCARD }
            val fields = mutableMapOf<FieldWrapper, FindFieldMethod>()
            val fieldKey = searchTerm.onlyClass('/')
            if (fieldKey == "*") {
                classes.forEach { (clazz, cm) ->
                    clazz.fields.forEach { field ->
                        if (fields.none { it.key.field == field })
                            fields[FieldWrapper(field, clazz, cm)] = FindFieldMethod.WILDCARD
                    }
                }
            } else {
                classes.forEach { (clazz, cm) ->
                    clazz.fields.forEach { field ->
                        if (fields.none { it.key.field == field })
                            when {
                                field.intermediaryName.contains(fieldKey, true) -> fields[FieldWrapper(field, clazz, cm)] = FindFieldMethod.INTERMEDIARY
                                field.mappedName?.contains(fieldKey, true) == true -> fields[FieldWrapper(field, clazz, cm)] = FindFieldMethod.MAPPED
                                field.obfName.client?.contains(fieldKey, true) == true -> fields[FieldWrapper(field, clazz, cm)] = FindFieldMethod.OBF_CLIENT
                                field.obfName.server?.contains(fieldKey, true) == true -> fields[FieldWrapper(field, clazz, cm)] = FindFieldMethod.OBF_SERVER
                                field.obfName.merged?.contains(fieldKey, true) == true -> fields[FieldWrapper(field, clazz, cm)] = FindFieldMethod.OBF_MERGED
                            }
                    }
                }
            }
            val sortedFields = when {
                fieldKey == "*" && (!hasClass || searchTerm.substring(0, searchTerm.lastIndexOf('/')).onlyClass() == "*") -> {
                    // Class and field both wildcard
                    fields.entries.sortedBy { it.key.field.intermediaryName }.sortedBy {
                        it.key.parent.mappedName?.onlyClass() ?: it.key.parent.intermediaryName
                    }.map { it.key }
                }
                fieldKey == "*" -> {
                    // Only field wildcard
                    val classKey = searchTerm.substring(0, searchTerm.lastIndexOf('/')).onlyClass()
                    fields.entries.sortedBy { it.key.field.intermediaryName }.sortedByDescending {
                        when (it.key.cm) {
                            FindFieldMethod.MAPPED -> it.key.parent.mappedName!!.onlyClass()
                            FindFieldMethod.OBF_CLIENT -> it.key.parent.obfName.client!!.onlyClass()
                            FindFieldMethod.OBF_SERVER -> it.key.parent.obfName.server!!.onlyClass()
                            FindFieldMethod.OBF_MERGED -> it.key.parent.obfName.merged!!.onlyClass()
                            FindFieldMethod.INTERMEDIARY -> it.key.parent.intermediaryName.onlyClass()
                            else -> null
                        }.similarityOnNull(classKey)
                    }.map { it.key }
                }
                hasClass && searchTerm.substring(0, searchTerm.lastIndexOf('/')).onlyClass() != "*" -> {
                    // has class
                    fields.entries.sortedByDescending {
                        when (it.value) {
                            FindFieldMethod.MAPPED -> it.key.field.mappedName!!
                            FindFieldMethod.OBF_CLIENT -> it.key.field.obfName.client!!
                            FindFieldMethod.OBF_SERVER -> it.key.field.obfName.server!!
                            FindFieldMethod.OBF_MERGED -> it.key.field.obfName.merged!!
                            else -> it.key.field.intermediaryName
                        }.onlyClass().similarity(fieldKey)
                    }.sortedBy {
                        it.key.parent.mappedName?.onlyClass() ?: it.key.parent.intermediaryName
                    }.map { it.key }
                }
                else -> {
                    fields.entries.sortedBy {
                        it.key.parent.mappedName?.onlyClass() ?: it.key.parent.intermediaryName
                    }.sortedByDescending {
                        when (it.value) {
                            FindFieldMethod.MAPPED -> it.key.field.mappedName!!
                            FindFieldMethod.OBF_CLIENT -> it.key.field.obfName.client!!
                            FindFieldMethod.OBF_SERVER -> it.key.field.obfName.server!!
                            FindFieldMethod.OBF_MERGED -> it.key.field.obfName.merged!!
                            else -> it.key.field.intermediaryName
                        }.onlyClass().similarity(fieldKey)
                    }.map { it.key }
                }
            }
            if (sortedFields.isEmpty())
                throw NullPointerException("No results found!")
            var page = 0
            val maxPage = ceil(sortedFields.size / 5.0).toInt()
            message.edit { it.setEmbed { it.buildMessage(sortedFields, mappingsContainer, page, user, maxPage) } }.subscribe { msg ->
                if (channel.type.name.startsWith("GUILD_"))
                    msg.removeAllReactions().block()
                msg.subscribeReactions("⬅", "❌", "➡")
                api.eventDispatcher.on(ReactionAddEvent::class.java).filter { e -> e.messageId == msg.id }.take(Duration.ofMinutes(15)).subscribe {
                    when (it.userId) {
                        api.selfId.get() -> {
                        }
                        user.id -> {
                            if (!it.emoji.asUnicodeEmoji().isPresent) {
                                msg.removeReaction(it.emoji, it.userId).subscribe()
                            } else {
                                val unicode = it.emoji.asUnicodeEmoji().get()
                                if (unicode.raw == "❌") {
                                    msg.delete().subscribe()
                                } else if (unicode.raw == "⬅") {
                                    msg.removeReaction(it.emoji, it.userId).subscribe()
                                    if (page > 0) {
                                        page--
                                        msg.edit { it.setEmbed { it.buildMessage(sortedFields, mappingsContainer, page, user, maxPage) } }.subscribe()
                                    }
                                } else if (unicode.raw == "➡") {
                                    msg.removeReaction(it.emoji, it.userId).subscribe()
                                    if (page < maxPage - 1) {
                                        page++
                                        msg.edit { it.setEmbed { it.buildMessage(sortedFields, mappingsContainer, page, user, maxPage) } }.subscribe()
                                    }
                                } else {
                                    msg.removeReaction(it.emoji, it.userId).subscribe()
                                }
                            }
                        }
                        else -> msg.removeReaction(it.emoji, it.userId).subscribe()
                    }
                }
            }
        } catch (t: Throwable) {
            try {
                message.edit { it.setEmbed { it.generateThrowable(t, user) } }.subscribe()
            } catch (throwable2: Throwable) {
                throwable2.addSuppressed(t)
                throw throwable2
            }
        }
    }
}

private enum class FindFieldMethod {
    INTERMEDIARY,
    MAPPED,
    OBF_CLIENT,
    OBF_SERVER,
    OBF_MERGED,
    WILDCARD
}

private data class FieldWrapper(val field: Field, val parent: Class, val cm: FindFieldMethod)

private fun EmbedCreateSpec.buildMessage(sortedMethods: List<FieldWrapper>, mappingsContainer: MappingsContainer, page: Int, author: User, maxPage: Int) {
    setFooter("Requested by " + author.discriminatedName, author.avatarUrl)
    setTimestampToNow()
    if (maxPage > 1) setTitle("List of ${mappingsContainer.name} Mappings (Page ${page + 1}/$maxPage)")
    var desc = ""
    sortedMethods.dropAndTake(5 * page, 5).forEach {
        if (!desc.isEmpty())
            desc += "\n\n"
        val obfMap = LinkedHashMap<String, String>()
        if (!it.field.obfName.isMerged()) {
            if (it.field.obfName.client != null) obfMap["client"] = it.field.obfName.client!!
            if (it.field.obfName.server != null) obfMap["server"] = it.field.obfName.server!!
        }
        desc += "**MC ${mappingsContainer.version}: ${it.parent.mappedName
                ?: it.parent.intermediaryName}.${it.field.mappedName ?: it.field.intermediaryName}**\n" +
                "__Name__: " + (if (it.field.obfName.isEmpty()) "" else if (it.field.obfName.isMerged()) "${it.field.obfName.merged} => " else "${obfMap.entries.joinToString { "${it.key}=**${it.value}**" }} => ") +
                "`${it.field.intermediaryName}`" + (if (it.field.mappedName == null || it.field.mappedName == it.field.intermediaryName) "" else " => `${it.field.mappedName}`")
        desc += "\n__Type__: `${(it.field.mappedDesc
                ?: it.field.intermediaryDesc.mapIntermediaryDescToNamed(mappingsContainer)).localiseFieldDesc()}`"
        desc += "\n__Mixin Target__: `L${it.parent.mappedName
                ?: it.parent.intermediaryName};${if (it.field.mappedName == null) it.field.intermediaryName else it.field.mappedName}:" +
                "${it.field.mappedDesc ?: it.field.intermediaryDesc.mapIntermediaryDescToNamed(mappingsContainer)}`"
    }
    setDescription(desc.substring(0, min(desc.length, 2000)))
}

private fun String.localiseFieldDesc(): String {
    if (length == 1) {
        val c = first()
        when (c) {
            'Z' -> return "boolean"
            'C' -> return "char"
            'B' -> return "byte"
            'S' -> return "short"
            'I' -> return "int"
            'F' -> return "float"
            'J' -> return "long"
            'D' -> return "double"
        }
    }
    if (length == 2 && first() == '[') {
        val c = last()
        when (c) {
            'Z' -> return "boolean[]"
            'C' -> return "char[]"
            'B' -> return "byte[]"
            'S' -> return "short[]"
            'I' -> return "int[]"
            'F' -> return "float[]"
            'J' -> return "long[]"
            'D' -> return "double[]"
        }
    }
    if (startsWith("[L"))
        return substring(2, length - 1).replace('/', '.') + "[]"
    if (startsWith("L"))
        return substring(1, length - 1).replace('/', '.')
    if (startsWith("[[L"))
        return substring(3, length - 1).replace('/', '.')
    return replace('/', '.')
}